test/extended/pods/oom_killer.go

```diff
@@ -0,0 +1,111 @@
+package pods
+
+import (
+	"context"
+	"fmt"
+	o "github.com/onsi/gomega"
+	"github.com/openshift/origin/pkg/test/ginkgo/result"
+	"strings"
+	"time"
+
+	g "github.com/onsi/ginkgo"
+	"github.com/openshift/library-go/pkg/build/naming"
+	exutil "github.com/openshift/origin/test/extended/util"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/uuid"
+	"k8s.io/kubernetes/test/e2e/framework"
+	e2epod "k8s.io/kubernetes/test/e2e/framework/pod"
+)
+
+var containerName = "stressng"
+
+var _ = g.Describe("[sig-node][Serial]", func() {
+	defer g.GinkgoRecover()
+
+	f := framework.NewDefaultFramework("validate-oom-kills")
+	oc := exutil.NewCLIWithFramework(f)
+
+	// The goal of this test to verify if the OOM inducing pod is the one that is getting OOM Killed
+	g.It("should oom kill the pod which induced it", func() {
+		currentTime := exutil.SuiteStartTime()
+		g.By("creating the OOM inducing pod")
+		podClient := f.PodClient()
+		ns := f.Namespace.Name
+		podName := naming.GetPodName("oom-inducing-pod", string(uuid.NewUUID()))
+		pod := &v1.Pod{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      podName,
+				Namespace: ns,
+			},
+			Spec: v1.PodSpec{
+				Containers: []v1.Container{{
+					Name:  containerName,
+					Image: "k8s.gcr.io/stress:v1",
+					Args:  []string{"--mem-alloc-size", "5Gi", "--mem-total", "25Gi"},
+				},
+				},
+				PriorityClassName: "system-cluster-critical",
+			},
+		}
+		pod = podClient.Create(pod)
+		err := e2epod.WaitForPodNameRunningInNamespace(oc.KubeFramework().ClientSet, pod.Name, oc.Namespace())
+		framework.ExpectNoError(err)
+
+		pod, err = podClient.Get(context.TODO(), pod.Name, metav1.GetOptions{})
+		o.Expect(err).NotTo(o.HaveOccurred())
+
+		defer func() {
+			g.By("deleting the pod")
+			podClient.Delete(context.TODO(), pod.Name, *metav1.NewDeleteOptions(0))
+		}()
+
+		// To wait for pod to start hogging memory
+		time.Sleep(2 * time.Minute)
+		oomEventMsgList := []string{}
+
+		// Collect the events to see if we're getting OOM Killed.
+		kubeClient := oc.AdminKubeClient()
+
+		events, err := kubeClient.CoreV1().Events("").List(context.TODO(), metav1.ListOptions{})
+		o.Expect(err).NotTo(o.HaveOccurred())
+		containerOOMKilledCount := 0 // We need to search for container exited with code 137
+		systemOOMErrorMsg := "System OOM encountered"
+		containerOOMErrorMsg := "container exited with code 137 (OOMKilled)"
```

Comment:
Explain to me why we don't get the same event again?  What is the difference between these two events, why do two different ones get sent, and why wouldn't we always get the system oom one?

Reply:
`System OOM events` are coming from oom_watcher.go which relies on cadvisor to parse kernel logs and return the output. `exit code 137` comes from container runtime which is bubbled up via kubelet.